Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs

```diff
@@ -4,31 +4,76 @@
     {
         private enum GazePointerState
         {
-            Initial, // When the application starts up, the gaze pointer should be active
-            GazePointerActive, // Gaze pointer is active when no hands are visible, after "select" 
-            GazePointerInactive // Gaze pointer is inactive as soon as motion controller or articulated hand pointers appear
+            // When the application starts up, the gaze pointer should be active
+            Initial,
+
+            // If head gaze is in use, then the gaze pointer is active when no hands are visible, after "select"
+            // If eye gaze is use, then the gaze pointer is active when no far pointers are active.
+            GazePointerActive,
+
+            // If head gaze is in use, then the gaze pointer is inactive as soon as motion controller or
+            // articulated hand pointers appear.
+            // If eye gaze is in use, then the gaze pointer is inactive when far pointers are active.
+            GazePointerInactive
         }
         private GazePointerState gazePointerState = GazePointerState.Initial;
         private bool activateGazeKeywordIsSet = false;
+        private bool eyeGazeValid = false;
 
         public bool IsGazePointerActive
         {
             get { return gazePointerState != GazePointerState.GazePointerInactive; }
         }
 
-        public void UpdateState(int numNearPointersActive, int numFarPointersActive)
+        /// <summary>
+        /// Updates the state machine based on the number of near pointers, the number of far pointers,
+        /// and whether or not eye gaze is valid.
+        /// </summary>
+        public void UpdateState(int numNearPointersActive, int numFarPointersActive, bool isEyeGazeValid)
+        {
+            if (eyeGazeValid != isEyeGazeValid)
+            {
+                activateGazeKeywordIsSet = false;
+                eyeGazeValid = isEyeGazeValid;
+            }
+
+            if (isEyeGazeValid)
+            {
+                UpdateStateEyeGaze(numNearPointersActive, numFarPointersActive);
+            }
+            else
+            {
+                UpdateStateHeadGaze(numNearPointersActive, numFarPointersActive);
+            }
+        }
+
+        private void UpdateStateEyeGaze(int numNearPointersActive, int numFarPointersActive)
+        {
+            // If there are any far pointers active while eye gaze is valid, then
+            // eye gaze should be disabled.
+            bool isEyeGazePointerActive = numFarPointersActive == 0;
+
+            gazePointerState = isEyeGazePointerActive ?
+                GazePointerState.GazePointerActive :
+                GazePointerState.GazePointerInactive;
+        }
+
+        private void UpdateStateHeadGaze(int numNearPointersActive, int numFarPointersActive)
         {
-            GazePointerState newState = gazePointerState;
             bool isMotionControllerOrHandUp = numFarPointersActive > 0 || numNearPointersActive > 0;
```

Comment:
This reminds me of bug #4405. Mind pulling that change in? It's very short, just replace with 

> bool canGazeCursorShow = numFarPointersActive == 1 && numNearPointersActive == 0;

Reply:
Done - I had to change up the suggested implementation a bit (i.e. #4405 had a bug in its proposed implementation in the missing negation of canGazeCursorShow within its if checks)